src/components/form/EmailInput.tsx

```diff
@@ -0,0 +1,84 @@
+import React from 'react';
+
+export interface IEmailProps {
+  warning?: boolean;
+  success?: boolean;
+  statusText?: string;
+  icon?: JSX.Element;
+  label: string;
+  register: any;
+  type?: string;
+}
+
+/**
+ *
+ * Email input form field component
+ *
+ * @param {boolean} warning - on validation failure
+ * @param {boolean} success - on validation success
+ * @param {string} statusText - validation status text
+ * @param {string} label - input label
+ *
+ * @returns {JSX.Element}
+ *
+ * @example
+ * <EmailInput success />
+ *
+ */
+function EmailInput({
+  warning,
+  success,
+  statusText,
+  label,
+  type,
+  icon,
+  register,
+  ...props
+}: IEmailProps) {
+  return (
+    <div className="my-8">
+      <div className="relative h-10">
+        {icon && <span className="absolute top-1/2 left-0 z-20 ml-2 -translate-y-1/2">{icon}</span>}
+
+        <input
+          type={type}
+          {...props}
+          name={label}
+          id={label}
+          {...register}
+          className={`absolute  border-2 rounded-sm py-2 ${
+            icon ? 'pl-8' : 'pl-2'
+          } pr-3 outline-cyan-100 peer focus:ring-0 ${
+            warning ? 'border-red-300 outline-red-300 ' : ''
+          }  ${success ? 'border-cyan-300 outline-cyan-300' : ''}`}
+          placeholder="  "
+        />
+        <label
+          htmlFor={label}
+          className={`[&:not(peer-placeholder-shown)]:text-dark [&:not(peer-placeholder-shown)]:translate-x-2 peer-placeholder-shown:translate-y-2  [&:not(peer-placeholder-shown)]:-translate-y-1/2 [&:not(peer-placeholder-shown)]:bg-white  peer-focus:bg-white peer-placeholder-shown:bg-white  peer-focus:-translate-y-3  peer-focus:translate-x-2 peer-placeholder-shown:translate-x-1 px-2 text-gray-300 peer-focus:text-black transition rounded-sm absolute ${
+            icon ? 'ml-4' : ''
+          } `}
+        >
+          {label}
+        </label>
```

Comment:
I think that is a good example but very limited, react form hooks can do so many things and that register method can have so many properties, I did it like that because you might want to do a lot of validations which can not be passed through props. This is an example of what I did 
```  
 <EmailInput
                key={2}
                label="Password"
                icon={<FaLock />}
                type="password"
                register={register('password', { required: true, pattern: /^(?=.*[a-z])(?=.*[A-Z])(?=.*?[0-9])(?=.*?[!#%])[A-Za-z0-9!#%]{8,32}$/ })}
                warning={!!errors.password}
              />
```

Reply:
Can't you pass optional params of those patterns to the component?


If it fails, you could only make one for a password, and leave the rest as shared.